model/DataObject.php

```diff
@@ -1417,7 +1418,12 @@ public function getComponent($componentName) {
 		if($class = $this->has_one($componentName)) {
 			$joinField = $componentName . 'ID';
 			$joinID    = $this->getField($joinField);
-			
+
+			if(isset(DataObject::$_cache_has_one[$class][$joinID])) {
+				$this->components[$componentName] = DataObject::$_cache_has_one[$class][$joinID];
+				return $this->components[$componentName];
+			}
+
 			if($joinID) {
 				$component = $this->model->$class->byID($joinID);
```

Comment:
We could use `DataObject::get_by_id($class, $joinID, true)` to use the caching in `DataObject::get_by_id`

I'm still a bit fuzzy on why we have DataModel. @chillu do you have a better idea?